File: back-end/01-ms-identity/src/main/java/com/faceattend_edu/identity_service/domain/model/UserSession.java
package com.faceattend_edu.identity_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.Duration;
import java.time.LocalDateTime;
import java.util.Objects;
import java.util.UUID;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class UserSession {
    private UUID sessionId;
    private User userId;
    private LocalDateTime startDate;
    private LocalDateTime endDate;
    private String sourceIp;
    private String sessionStatus;

    public void start() {
        if (this.sessionId == null) this.sessionId = UUID.randomUUID();
        this.startDate = LocalDateTime.now();
        this.endDate = null;
        this.sessionStatus = "ACTIVE";
    }

    public void end() {
        this.endDate = LocalDateTime.now();
        this.sessionStatus = "ENDED";
    }

    public boolean isActive() {
        return "ACTIVE".equals(this.sessionStatus) && this.endDate == null;
    }

    public long durationSeconds() {
        if (this.startDate == null) return 0L;
        LocalDateTime end = this.endDate == null ? LocalDateTime.now() : this.endDate;
        return Math.max(0, Duration.between(this.startDate, end).getSeconds());
    }
}

File: front-end/Mobile/src/models/identity/UserSession.js
import BaseModel from '../BaseModel';

export default class UserSession extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.sessionId = data.session_id || null;
    this.userId = data.user_id || null;
    this.startDate = data.start_date || null;
    this.endDate = data.end_date || null;
    this.sourceIp = data.source_ip || null;
    this.sessionStatus = data.session_status || 'Active';
  }

  get isActive() {
    return this.sessionStatus === 'Active';
  }

  get duration() {
    if (!this.startDate) return null;
    const end = this.endDate ? new Date(this.endDate) : new Date();
    return end - new Date(this.startDate);
  }

  static fromApi(data) {
    if (!data) return null;
    return new UserSession(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      session_id: this.sessionId,
      user_id: this.userId,
      start_date: this.startDate,
      end_date: this.endDate,
      source_ip: this.sourceIp,
      session_status: this.sessionStatus,
    };
  }
}
